src/Orleans/IDs/SiloAddress.cs

```diff
@@ -253,7 +253,48 @@ public bool Equals(SiloAddress other)
 
         public int CompareTo(SiloAddress other)
         {
-            return other == null ? 1 : Generation.CompareTo(other.Generation);
+            if (other == null) return 1;
+            // Compare Generation first. It gives a cheap and fast way to compare, avoiding allocations 
+            // and is also semantically meaningfull - older silos (with smaller Generation) will appear first in the comparison order.
+            // Only if Generations are the same, go on to compare Ports and IPAddress (which is more expansive to compare).
+            // Alternatively, we could compare ConsistentHashCode or UniformHashCode.
+            int comp = Generation.CompareTo(other.Generation);
+            if (comp != 0) return comp;
+
+            comp = Endpoint.Port.CompareTo(other.Endpoint.Port);
+            if (comp != 0) return comp;
+
+            return CompareIpAddresses(Endpoint.Address, other.Endpoint.Address);
+        }
+
+        // Taken from: http://www.codeproject.com/Articles/26550/Extending-the-IPAddress-object-to-allow-relative-c
+        private static int CompareIpAddresses(IPAddress one, IPAddress two)
+        {
+            int returnVal = 0;
+            if (one.AddressFamily == two.AddressFamily)
+            {
+                byte[] b1 = one.GetAddressBytes();
+                byte[] b2 = two.GetAddressBytes();
```

Comment:
I wonder should we treat, say, `127.0.0.1` and `::1` same here? This isn't likely just this straightforward or then `GetAddressBytes` does more than I'd imagine. I haven't checked, though. What I mean is that there are functions that map IPv4 and IPv6 addresses to each other, like at http://stackoverflow.com/questions/16618810/compare-ipv4-addresses-in-ipv6-notation or that one doesn't need to enter all octets as described at http://stackoverflow.com/questions/22187690/ip-address-class-getaddressbytes-method-putting-octets-in-odd-indices-of-the-byt.

If nothing else, it might be good to record these concerns. A tangential note could be made towards dual stack sockets, described briefly at https://msdn.microsoft.com/en-us/library/system.net.ipaddress.maptoipv6(v=vs.110).aspx.